x-pack/plugin/sql/qa/mixed-node/src/test/java/org/elasticsearch/xpack/sql/qa/mixed_node/SqlCompatIT.java

```diff
@@ -154,42 +168,109 @@ public static String sqlQueryEntityWithOptionalMode(String query, Version bwcVer
+    }
+
+    private void assertCursorCloseWorksAcrossVersions(Version version1, RestClient client1, RestClient client2) throws IOException {
+        indexDocs();
+
+        String cursor = fetch1RecordAndReturnCursor(version1, client1);
+
+        Request scrollReq = new Request("POST", "_sql/close");
+        scrollReq.addParameter("error_trace", "true");
+        scrollReq.setJsonEntity("{\"cursor\": \"%s\"}".formatted(cursor));
+        Map<String, Object> scrollJson = performRequestAndReadBodyAsJson(client2, scrollReq);
+
+        assertEquals(scrollJson.get("succeeded"), true);
+    }
+
+    public void testCursorForIndexAllocatedOnNewNodeWorkOnOldNode() throws IOException {
+        assumeTrue(
+            "requires all nodes supporting cursor redirection in case of version mismatch",
+            bwcVersion.onOrAfter(INTRODUCE_REDIRECTS_FOR_CURSORS_WITH_VERSION_MISMATCH_VERSION)
+        );
+
+        indexDocs(Map.of("index.number_of_shards", 1, "index.number_of_replicas", 0, "index.routing.allocation.require.upgraded", "true"));
+
+        String cursor = fetch1RecordAndReturnCursor(Version.CURRENT, newNodesClient);
+        assertCursorCanFetchDocs(oldNodesClient, cursor);
+    }
+
+    private static String fetch1RecordAndReturnCursor(Version version, RestClient client) throws IOException {
         Request req = new Request("POST", "_sql");
-        // GROUP BY queries always return a cursor
-        req.setJsonEntity(sqlQueryEntityWithOptionalMode("SELECT int FROM test GROUP BY 1", bwcVersion));
-        Map<String, Object> json = performRequestAndReadBodyAsJson(client1, req);
+        req.setJsonEntity(sqlQueryEntityWithOptionalMode("SELECT int FROM test", version, 1));
+        Map<String, Object> json = performRequestAndReadBodyAsJson(client, req);
         String cursor = (String) json.get("cursor");
         assertThat(cursor, Matchers.not(Matchers.emptyString()));
+        return cursor;
+    }
 
+    private static void assertCursorCanFetchDocs(RestClient client, String cursor) throws IOException {
         Request scrollReq = new Request("POST", "_sql");
+        scrollReq.addParameter("error_trace", "true");
         scrollReq.setJsonEntity("{\"cursor\": \"%s\"}".formatted(cursor));
-        ResponseException exception = expectThrows(ResponseException.class, () -> client2.performRequest(scrollReq));
+        Map<String, Object> scrollJson = performRequestAndReadBodyAsJson(client, scrollReq);
 
-        assertThat(
-            exception.getMessage(),
-            Matchers.containsString("Unsupported cursor version [" + version1 + "], expected [" + version2 + "]")
-        );
+        assertNotNull(scrollJson.get("rows"));
+    }
+
+    public void testTextCursorFromOldNodeWorksOnNewNode() throws IOException {
+        assertTextCursorCompatibleAcrossVersions(bwcVersion, oldNodesClient, newNodesClient);
+    }
+
+    @AwaitsFix(bugUrl = "tbd")
```

Comment:
See last remark in the PR description. I'll create an according issue once this PR is approved and ready to merge.

Reply:
Never mind about this comment. The `@AwaitsFix` referred to a previous attempt to address the issue. With the current attempt, text format based cursors should also be usable across versions if all nodes support the cursor redirect from this PR.